File: academic-system-java/src/main/java/org/example/academic/system/security/AuthenticationService.java
package org.example.academic.system.security;

import org.example.academic.system.exception.AuthenticationException;
import org.example.academic.system.model.User;
import org.example.academic.system.repository.UserRepository;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.Optional;

public class AuthenticationService {

    private static final Logger logger = LoggerFactory.getLogger(AuthenticationService.class);

    private final UserRepository userRepository;
    private final Session session = Session.getInstance();

    public AuthenticationService(UserRepository userRepository) {
        this.userRepository = userRepository;
    }

    public User authenticate(String username, String password) {
        Optional<User> userOptional = userRepository.findByUsername(username);

        if (userOptional.isEmpty() || !userOptional.get().getPassword().equals(password)) {
            logger.warn("Failed authentication attempt for username: {}", username);
            throw new AuthenticationException("Invalid credentials");
        }

        User user = userOptional.get();
        session.login(user);
        logger.info("User '{}' authenticated successfully with role: {}", username, user.getRole());
        return user;
    }

    public void logout() {
        User user = session.getAuthenticatedUser();
        if (user != null) {
            logger.info("User '{}' logged out", user.getUsername());
        }
        session.logout();
    }
}

File: academic-system-python/academic_system/security/authentication_service.py
import logging

from ..exception.authentication_exception import AuthenticationException
from ..model.user import User
from ..repository.user_repository import UserRepository
from .session import Session

logger = logging.getLogger(__name__)


class AuthenticationService:
    def __init__(self, user_repository: UserRepository):
        self._user_repository = user_repository
        self._session = Session.get_instance()

    # Valida as credenciais e abre a sessão se corretas; lança exceção se inválidas
    def authenticate(self, username: str, password: str) -> User:
        user = self._user_repository.find_by_username(username)

        if user is None or user.password != password:
            logger.warning("Failed authentication attempt for username: %s", username)
            raise AuthenticationException("Invalid credentials")

        self._session.login(user)
        logger.info("User '%s' authenticated successfully with role: %s", username, user.role)
        return user

    # Encerra a sessão do usuário autenticado, limpando o estado da sessão
    def logout(self) -> None:
        user = self._session.get_authenticated_user()
        if user:
            logger.info("User '%s' logged out", user.username)
        self._session.logout()
